components/datetime/src/pattern/item/ule.rs

```diff
@@ -0,0 +1,215 @@
+///
+/// The serialization model packages the pattern item in three bytes.
+///
+/// The first bit is used to disriminate the item variant. If the bit is
+/// set, then the value is the `PatternItem::Field` variant. Otherwise,
+/// the `PatternItem::Literal` is used.
+///
+/// In case the discriminant is set:
+///
+/// 1) The rest of the first byte remains usnused.
+/// 2) The second byte encodes `FieldSymbol` encoded as (Type: 4 bits, Symbol: 4 bits).
+/// 3) The third byte encodes the field length.
+///
+/// If the discriminant is not set, the bottom three bits of the first byte,
+/// together with the next two bytes, contain all 21 bits required to encode
+/// any [`Unicode Code Point`].
+///
+/// # Diagram
+///
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │       u8      │       u8      │       u8      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┤
+/// ├─┴─┴─┼─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┤
+/// │     │          Unicode Code Point             │ Literal
+/// ├─┬───┴─────────┬───────────────┬───────────────┤
+/// │X│             │  FieldSymbol  │  FieldLength  │ Field
+/// └─┴─────────────┴───────────────┴───────────────┘
+///  ▲
+///  │
+///  Variant Discriminant
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `PatternItem` elements
+/// and performant deserialization from the `PatternItemULE` to `PatternItem` type.
+///
+/// # Constraints
+///
+/// The model leaves at most 8 `PatternItem` variants, limits the number of possible
+/// field types and symbols to 16 each and limits the number of length variants to 128.
+///
+/// [`Unicode Code Point`]: http://www.unicode.org/versions/latest/
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct PatternItemULE([u8; 3]);
+
+impl PatternItemULE {
+    #[inline]
+    fn determine_field_from_u8(byte: u8) -> bool {
+        byte & 0b1000_0000 != 0
+    }
+
+    #[inline]
+    fn bytes_in_range(value: (&u8, &u8, &u8)) -> bool {
+        match Self::determine_field_from_u8(*value.0) {
+            true => fields::Field::bytes_in_range(value.1, value.2),
+            false => {
+                let u = u32::from_be_bytes([0x00, *value.0, *value.1, *value.2]);
+                char::try_from(u).is_ok()
+            }
+        }
+    }
+}
+
+unsafe impl ULE for PatternItemULE {
+    type Error = &'static str;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let mut chunks = bytes.chunks_exact(3);
+
+        if !chunks.all(|c| Self::bytes_in_range((&c[0], &c[1], &c[2])))
+            || !chunks.remainder().is_empty()
+        {
+            return Err("Invalid byte sequence");
+        }
+        Ok(())
+    }
+}
+
+impl AsULE for PatternItem {
+    type ULE = PatternItemULE;
+
+    #[inline]
+    fn as_unaligned(&self) -> Self::ULE {
+        match self {
+            Self::Field(field) => {
+                PatternItemULE([0b1000_0000, field.symbol.idx(), field.length as u8])
```

Comment:
Nit:

```suggestion
                PatternItemULE([0b1000_0000, field.symbol.idx(), field.length.idx()])
```

for consistency and keeping things centralized.